okhttp/src/main/java/okhttp3/internal/http/HttpEngine.java

```diff
@@ -356,17 +293,13 @@ public Connection getConnection() {
    * engine that should be used for the retry if {@code e} is recoverable, or null if the failure is
    * permanent. Requests with a body can only be recovered if the body is buffered.
    */
-  public HttpEngine recover(IOException e, boolean routeException, Sink requestBodyOut) {
+  public HttpEngine recover(IOException e, boolean routeException) {
     streamAllocation.streamFailed(e);
 
     if (!client.retryOnConnectionFailure()) {
       return null; // The application layer has forbidden retries.
     }
 
-    if (requestBodyOut != null && !(requestBodyOut instanceof RetryableSink)) {
```

Comment:
Here's what I mean, it may not be valid, although it does seem to pass with 3.3.1 and 3.2. If it is valid, I think it happened during the HttpURLConnection -> native transition.

```java
  @Test public void streamedBodyNotRepeated() throws Exception {
    server.enqueue(new MockResponse().setSocketPolicy(SocketPolicy.DISCONNECT_AFTER_REQUEST));

    Dns dns = new FakeDns().addresses(Arrays.asList(InetAddress.getByName("localhost"),
        InetAddress.getByName("localhost")));
    OkHttpClient client = new OkHttpClient.Builder()
        .dns(dns)
        .build();
    factory = new OkUrlFactory(client);
    HttpURLConnection connection = factory.open(server.url("/").url());
    connection.setDoOutput(true);
    connection.setChunkedStreamingMode(100);
    OutputStream os = connection.getOutputStream();
    os.write("OutputStream is no fun.".getBytes("UTF-8"));
    os.close();

    try {
      connection.getResponseCode();
      fail();
    } catch (IOException expected) {
    }

    assertEquals(1, server.getRequestCount());
  }
```